### PR TITLE
checkstyle: check for unused import #252

### LINKED ISSUE #252: checkstyle: check for unused import
Currently, unused import are not checked when running `checkstyleMain` or `checkstyleTest`, though codacy will check that.

We can reproduce by adding `import java.util.Arrays;` to `MainApp.java` and run `checkstyleMain`, verify it reports BUILD SUCCESSFUL.

### PR DESCRIPTION
**Pull request by chao1995**
Fixes #252 

Ready for review.

### CODE REVIEW

**Comment by pyokagan** [comment]
@chao1995 
> Ready for review.

Please use @CanIHasReview

**Comment by CanIHasReview** [comment]
# v1

@chao1995 submitted v1 for review.

* [v1 1/1] [Add UnusedImports check in checkstyle](https://github.com/se-edu/addressbook-level4/pull/256/commits/ac993c084662f8ccb6f1170d87824137f0ff59cb)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/8164678f10e07049144c98acd52fc3572ad5c9a8...ac993c084662f8ccb6f1170d87824137f0ff59cb))
<details>
<summary>Checkout this PR version locally</summary>

```shell
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/256/1/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you want to fetch this PR to.

</details>

**Comment by pyokagan** [comment]
@chao1995 Please request at least one peer reviewer for this PR.

File: config/checkstyle/checkstyle.xml [hunk 1]
```diff
@@ -44,7 +44,20 @@
     -->
 
     <module name="RedundantImport">
-      <!-- Checks for redundant import statements. -->
+      <!-- Checks for redundant import statements.
+      An import statement is redundant if:
+        It is a duplicate of another import. This is, when a class is imported more than once.
```
Comment (yamgent) [inline config/checkstyle/checkstyle.xml]: You might want to add bullet points here, wasn't clear to me that it was supposed to be a list when I first read it. :P

File: config/checkstyle/checkstyle.xml [hunk 2]
```diff
@@ -44,7 +44,20 @@
...
+        The class non-statically imported is from the java.lang package, e.g. importing java.lang.String.
+        The class non-statically imported is from the same package as the current package.
```
Comment (yamgent) [inline config/checkstyle/checkstyle.xml]: `The class **is** non-statically imported`

**Comment by CanIHasReview** [comment]
# v2

@chao1995 submitted v2 for review.

* [v2 1/1] [Add UnusedImports check in checkstyle](https://github.com/se-edu/addressbook-level4/pull/256/commits/8bc89cafbb71612ff7fd99e2ffac68ae1393419b)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/f74c40cb73cd505ba1d6540bd9361cab1369d830...8bc89cafbb71612ff7fd99e2ffac68ae1393419b)) ([:chart_with_upwards_trend: Interdiff between v1 and v2](https://github.com/se-edu/addressbook-level4/commit/786b0d6c8e31b83a97968514f8451d7ff66263f7))
<details>
<summary>Checkout this PR version locally</summary>

```shell
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/256/2/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you want to fetch this PR to.

</details>

File: config/checkstyle/checkstyle.xml (hunk 1 above)
Comment (chao1995) [inline config/checkstyle/checkstyle.xml]: @yamgent Thanks for pointing out! Added bullet points.

File: config/checkstyle/checkstyle.xml (hunk 2 above)
Comment (chao1995) [inline config/checkstyle/checkstyle.xml]: Thanks for the suggestion. Both positions for the "is" are syntactically correct, maybe we can keep this one same as checkstyle's [documentation](http://checkstyle.sourceforge.net/config_imports.html#RedundantImport#RedundantImport).

**Comment by yamgent** [comment]
Cosmetic change in the commit message:

> This commit included UnusedImports to check for unused imports and
> added explanation about the difference between these two checks.

When talking about changes done in the current, we need to use imperative language (like we are giving orders to the codebase). So it should say something like `Therefore, include UnusedImports to check for unused imports, and add explanation...`

**Comment by damithc** [comment]
> When talking about changes done in the current, we need to use imperative language (like we are giving orders to the codebase)

Confirmed must use imperative language for commit message title. I don't think the same applies to commit message body though, based on examples in https://github.com/git/git/commits/master?author=peff . But we should use present tense to explain the current commit. e.g. `this commit included` -> `this commit includes`

**Comment by CanIHasReview** [comment]
# v3

@chao1995 submitted v3 for review.

* [v3 1/1] [Add UnusedImports check in checkstyle](https://github.com/se-edu/addressbook-level4/pull/256/commits/fa6ca99f3a239d3191ff3e63cdb64f232a788355)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/f74c40cb73cd505ba1d6540bd9361cab1369d830...fa6ca99f3a239d3191ff3e63cdb64f232a788355)) ([:chart_with_upwards_trend: Interdiff between v2 and v3](https://github.com/se-edu/addressbook-level4/commit/971efdacdbcd5bf46919874a8ec75647ebaf78e2))
<details>
<summary>Checkout this PR version locally</summary>

```shell
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/256/3/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you want to fetch this PR to.

</details>

**Comment by chao1995** [comment]
@yamgent @damithc Fixed commit message as discussed.

**Comment by yamgent** [comment]
@chao1995 I don't think we are allowed to change the labels to `s.MergeApproved` until either Paul or Prof reviews the PR as well. :P

**Comment by chao1995** [comment]
@pyokagan @damithc Any further changes for this PR?

**Comment by pyokagan** [comment]
@damithc 
> I don't think the same applies to commit message body though, based on examples in https://github.com/git/git/commits/master?author=peff

What examples? It's likely just unintended. I selected this author because he seems to consistently use present tense to describe the code base *before* the commit is applied, and imperative form to describe the changes done by the commit.

> But we should use present tense to explain the current commit. e.g. this commit included -> this commit includes

I'd like us to avoid writing `This commit includes`, it adds bloat to the commit message since readers already know they are reading a commit. Instead, use the shorter and more direct imperative form, e.g. `Use UnusedImports to check for unused imports and add an explanation about the difference between these two checks.`

**Comment by damithc** [comment]
@pyokagan thanks for the correction. So let's stick with a format like this:

```txt
Replace Foo with Bar

Foo is slow. It is also hard to understand.

Change it to Bar so it can be faster. Do the migration
for existing data at the same time.

Don't use Goo because it is not free.
```

That is,
* Subject: in imperative form
* Body
  * Describe current state in present tense, but avoid terms such as 'currently, previously, etc.'
  * Describe commit (what is being done, and why) in imperative form, but avoid dead wood phrases such `this commit`

Anything else? Once settled, I'll update the docs.

File: config/checkstyle/checkstyle.xml [hunk 3]
```diff
@@ -44,7 +44,20 @@
...
 
     <module name="RedundantImport">
-      <!-- Checks for redundant import statements. -->
+      <!-- Checks for redundant import statements.
+      An import statement is redundant if:
+        * It is a duplicate of another import. This is, when a class is imported more than once.
+        * The class non-statically imported is from the java.lang package, e.g. importing java.lang.String.
+        * The class non-statically imported is from the same package as the current package.
+      -->
+      <property name="severity" value="error"/>
```
Comment (pyokagan) [inline config/checkstyle/checkstyle.xml]: We don't need this, right? checkstyle still gives me errors without this line.

Comment (pyokagan) [inline config/checkstyle/checkstyle.xml]: The reason is likely because the default checkstyle severity is "error".

**Comment by CanIHasReview** [comment]
# v4

@chao1995 submitted v4 for review.

* [v4 1/1] [Add UnusedImports check in checkstyle](https://github.com/se-edu/addressbook-level4/pull/256/commits/0940bf947c94bbdfb737cea191c2bd1c60a0ba10)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/f74c40cb73cd505ba1d6540bd9361cab1369d830...0940bf947c94bbdfb737cea191c2bd1c60a0ba10)) ([:chart_with_upwards_trend: Interdiff between v3 and v4](https://github.com/se-edu/addressbook-level4/commit/89a19b7dfb151d3ea479291c5bb082de0adb084b))
<details>
<summary>Checkout this PR version locally</summary>

```shell
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/256/4/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you want to fetch this PR to.

</details>

**Pull request merged**